ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -396,6 +368,225 @@ object Queries {
+      trackMatchIndices: MatchedQueryMarker[T, Mark],
+  )(implicit
+      log: LogHandler,
+      gvs: Get[Vector[String]],
+      pvs: Put[Vector[String]],
+  ): T[Query0[DBContract[Mark, JsValue, JsValue, Vector[String]]]] = {
+    val partyVector = parties.toVector
+    def query(preds: OneAnd[Vector, (SurrogateTpId, Fragment)], findMark: SurrogateTpId => Mark) = {
+      val assocedPreds = preds.map { case (tpid, predicate) =>
+        sql"(tpid = $tpid AND (" ++ predicate ++ sql"))"
+      }
+      val unionPred = concatFragment(intersperse(assocedPreds, sql" OR "))
+      val q = sql"""SELECT contract_id, tpid, key, payload, signatories, observers, agreement_text
+                      FROM contract AS c
+                      WHERE (signatories && $partyVector::text[] OR observers && $partyVector::text[])
+                       AND (""" ++ unionPred ++ sql")"
+      q.query[(String, SurrogateTpId, JsValue, JsValue, Vector[String], Vector[String], String)]
+        .map { case (cid, tpid, key, payload, signatories, observers, agreement) =>
+          DBContract(
+            contractId = cid,
+            templateId = findMark(tpid),
+            key = key,
+            payload = payload,
+            signatories = signatories,
+            observers = observers,
+            agreementText = agreement,
+          )
+        }
+    }
+
+    trackMatchIndices match {
+      case MatchedQueryMarker.ByInt =>
+        type Ix = Int
+        uniqueSets(queries.zipWithIndex map { case ((tpid, pred), ix) => (tpid, (pred, ix)) }).map {
+          preds: Map[SurrogateTpId, (Fragment, Ix)] =>
+            val predHd +: predTl = preds.toVector
+            val predsList = OneAnd(predHd, predTl).map { case (tpid, (predicate, _)) =>
+              (tpid, predicate)
+            }
+            query(predsList, tpid => preds(tpid)._2)
+        }
+
+      case MatchedQueryMarker.Unused =>
+        val predHd +: predTl = queries.toVector
+        query(OneAnd(predHd, predTl), identity)
+    }
+  }
 }
 
-private object PostgresQueries extends Queries
+private object OracleQueries extends Queries {
+  import Queries.{DBContract, MatchedQueryMarker, SurrogateTpId}, Queries.InitDdl.CreateTable
+  import Implicits._
+
+  protected[this] override def dropTableIfExists(table: String) = sql"""BEGIN
+      EXECUTE IMMEDIATE 'DROP TABLE ' || $table;
+    EXCEPTION
+      WHEN OTHERS THEN
+        IF SQLCODE != -942 THEN
+          RAISE;
+        END IF;
+    END;"""
+
+  protected[this] override def bigIntType = sql"NUMBER(19,0)"
+  protected[this] override def bigSerialType =
+    bigIntType ++ sql" GENERATED by default on null as IDENTITY"
+  protected[this] override def textType = sql"NVARCHAR2(100)"
+  protected[this] override def agreementTextType = sql"NVARCHAR2(100)"
+
+  protected[this] override def jsonColumn(name: Fragment) =
+    name ++ sql" CLOB CONSTRAINT ensure_json_" ++ name ++ sql" CHECK (" ++ name ++ sql" IS JSON)"
```

Comment:
Did you consider the implications of CLOB vs NCLOB vs UTF8 here? Probably worth documenting why you chose CLOB in a comment.